subprojects/execution/src/main/java/org/gradle/internal/execution/ExecutionEngine.java

```diff
@@ -30,28 +30,56 @@
 import java.util.function.Supplier;
 
 public interface ExecutionEngine {
-    /**
-     * Execute the given unit of work using available optimizations like
-     * up-to-date checks, build cache and incremental execution.
-     */
-    Result execute(UnitOfWork work);
-
-    /**
-     * Force the re-execution of the given unit of work disabling optimizations
-     * like up-to-date checks, build cache and incremental execution.
-     *
-     * @param reason the reason to report for rebuilding the given unit of work.
-     */
-    Result rebuild(UnitOfWork work, String reason);
-
-    /**
-     * Load the given unit from the given cache, or defer its execution.
-     *
-     * If the cache already contains the outputs for the given work, it is passed directly to {@link DeferredExecutionHandler#processCachedOutput(Try)}.
-     * Otherwise the execution is wrapped in deferred via {@link DeferredExecutionHandler#processDeferredOutput(Supplier)}.
-     * The work is looked up by its {@link UnitOfWork.Identity identity} in the given cache.
-     */
-    <T, O> T getFromIdentityCacheOrDeferExecution(UnitOfWork work, Cache<Identity, Try<O>> cache, DeferredExecutionHandler<O, T> handler);
+    DirectExecutionRequestBuilder createRequest(UnitOfWork work);
+
+    interface Builder {
```

Comment:
```suggestion
    interface Request {
```

I would call the builder `Request`. It doesn't really look like a builder, so I think request is a good name for this thing. `RequestBuilder` would also be fine for me, though the extra `Builder` doesn't add much I think.

Reply:
I think I managed to simplify this significantly now.